File: python/signal/freqswap.cpp
/* Copyright (c) 2009 by CodeSourcery.  All rights reserved.

   This file is available for license from CodeSourcery, Inc. under the terms
   of a commercial license and under the GPL.  It is not part of the VSIPL++
   reference implementation and is not available under the BSD license.
*/
/** @file    python/signal/freqswap.cpp
    @author  Stefan Seefeld
    @date    2009-09-11
    @brief   VSIPL++ Library: Python bindings for signal module.

*/
#include <boost/python.hpp>
#include <boost/noncopyable.hpp>
#include <vsip/signal.hpp>
#include <vsip/vector.hpp>
#include <vsip/matrix.hpp>
#include "../block.hpp"

namespace pyvsip
{

template <typename T>
vsip::Vector<T, Block<1, T> >
freqswap1(vsip::Vector<T, Block<1, T> > in) { return vsip::freqswap(in);}

template <typename T>
vsip::Matrix<T, Block<2, T> >
freqswap2(vsip::Matrix<T, Block<2, T> > in)
{ return vsip::freqswap(in);}
}

BOOST_PYTHON_MODULE(freqswap)
{
  using namespace pyvsip;

  bpl::def("freqswap", freqswap1<float>);
  bpl::def("freqswap", freqswap1<double>);
  bpl::def("freqswap", freqswap2<float>);
  bpl::def("freqswap", freqswap2<double>);
}

File: python/src/signal/freqswap.py
#
# Copyright (c) 2014 Stefan Seefeld
# All rights reserved.
#
# This file is part of OpenVSIP. It is made available under the
# license contained in the accompanying LICENSE.BSD file.

import numpy
from vsip import vector, matrix

def _import_module(dtype):
    mod = None
    if dtype == numpy.float32:
        _temp = __import__('vsip.signal', globals(), locals(), ['_freqswap_f'], -1) 
        mod = _temp._freqswap_f
    elif dtype in (float, numpy.float64):
        _temp = __import__('vsip.signal', globals(), locals(), ['_freqswap_d'], -1) 
        mod = _temp._freqswap_d
    if not mod:
        raise ValueError, 'Unsupported dtype %s'%(dtype)
    return mod


def freqswap(v):

    m = _import_module(v.dtype)
    if isinstance(v, vector):
        vout = vector(v.dtype, v.length())
    else:
        vout = matrix(v.dtype, v.rows(), v.cols())
    m.freqswap(v.block, vout.block)
    return vout
